### PR TITLE
Support evaluation of predicates within shortestPath

### PR DESCRIPTION
**Pull request by craigtaverner**
- All predicates evaluated on each path found before shortest filtering
- ALL(...) and NONE(...) predicates evaluated within Expander for early branch pruning performance

### CODE REVIEW

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/ShortestPathExpression.scala [hunk 1]
```diff
@@ -45,51 +49,109 @@ case class ShortestPathExpression(ast: ShortestPath) extends Expression with Pat
...
-    new SingleShortestPathStrategy(expander, ast.allowZeroLength, ast.maxDepth.getOrElse(15))
+  private def addPredicates(ctx: ExecutionContext, relTypeAndDirExpander: Expander)(implicit state: QueryState): Expander = if (predicates.isEmpty) relTypeAndDirExpander
   else
-    new AllShortestPathsStrategy(expander, ast.allowZeroLength, ast.maxDepth.getOrElse(15))
-
-  def calculateType(symbols: SymbolTable) =  shortestPathStrategy.typ
+    predicates.foldLeft(relTypeAndDirExpander) {
+      case (currentExpander, predicate) =>
+        predicate match {
+          case NoneInCollection(RelationshipFunction(_), symbolName, innerPredicate) => addAllOrNoneRelationshipExpander(ctx, currentExpander, false, innerPredicate, symbolName)
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/ShortestPathExpression.scala]: name boolean parameter

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/ShortestPathExpression.scala [hunk 2]
```diff
@@ -45,51 +49,109 @@ case class ShortestPathExpression(ast: ShortestPath) extends Expression with Pat
...
+          case AllInCollection(RelationshipFunction(_), symbolName, innerPredicate) => addAllOrNoneRelationshipExpander(ctx, currentExpander, true, innerPredicate, symbolName)
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/ShortestPathExpression.scala]: name boolean parameter

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/ShortestPathExpression.scala [hunk 3]
```diff
@@ -45,51 +49,109 @@ case class ShortestPathExpression(ast: ShortestPath) extends Expression with Pat
...
 
-  private lazy val expander: Expander = if (ast.relTypes.isEmpty) {
-    Traversal.expanderForAllTypes(toGraphDb(ast.dir))
+  private def cypherPredicatesAsExpander(incomingCtx: ExecutionContext, relName: String, predicate: Predicate)(implicit state: QueryState) = new org.neo4j.function.Predicate[Relationship] {
+    override def test(t: Relationship): Boolean = {
+      predicate.isTrue(incomingCtx.newWith(relName -> t))
+    }
+  }
+
+  private def addAllOrNoneRelationshipExpander(ctx: ExecutionContext, currentExpander: Expander, all: Boolean, predicate: Predicate, relName: String)(implicit state: QueryState) = {
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/ShortestPathExpression.scala:122]: Can't you have different functions for `all == true` and `all = false`

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/ShortestPathExpression.scala [hunk 4]
```diff
@@ -45,51 +49,109 @@ case class ShortestPathExpression(ast: ShortestPath) extends Expression with Pat
...
+      case PropertyExists(_, propertyKey) =>
+        currentExpander.addRelationshipFilter(
+          if(all) propertyExistsExpander(propertyKey.name)
+          else propertyNotExistsExpander(propertyKey.name))
+      case Not(PropertyExists(_, propertyKey)) =>
+        currentExpander.addRelationshipFilter(
+          if(all) propertyNotExistsExpander(propertyKey.name)
+          else propertyExistsExpander(propertyKey.name))
+      case _ => currentExpander.addRelationshipFilter(
+        cypherPredicatesAsExpander(ctx, relName, predicate))
```
Comment (craigtaverner) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/ShortestPathExpression.scala]: Why no if(all)? Looks like a mistake here.

**Pull request merged**